File: algorithms/105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.py
"""
105. 从前序与中序遍历序列构造二叉树
"""

from logging import root
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def build_tree(preorder: List[int], inorder: List[int]) -> TreeNode:
    if not preorder:
        return None
    root = TreeNode(preorder[0])
    i = 0
    while i < len(inorder):
        if inorder[i] == preorder[0]:
            break
        else:
            i += 1
    # 递归构建
    left_count = i
    root.left = build_tree(preorder[1:left_count+1], inorder[:i])
    root.right = build_tree(preorder[left_count+1:], inorder[i+1:])
    return root

File: algorithms/105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.go
package problem105

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

// 递归。递归构造树以及左右子树
func buildTree(preorder []int, inorder []int) *TreeNode {
	if len(preorder) == 0 {
		return nil
	}

	// 根据前序遍历特点，先找到根节点
	root := &TreeNode{preorder[0], nil, nil}
	i := 0
	for ; i < len(inorder); i++ {
		if inorder[i] == preorder[0] {
			break
		}
	}

	// 递归构建，要分别知道左子树和右子树节点的数量
	leftCount := len(inorder[:i])
	root.Left = buildTree(preorder[1:leftCount+1], inorder[:i])
	root.Right = buildTree(preorder[leftCount+1:], inorder[i+1:])
	return root
}